### ISSUE TITLE
Unsafe usage of `std::slice::from_raw_parts`

### ISSUE THREAD

**Issue by gautamg795** [qdrant/qdrant#3802]
`segment::common::mmap_ops::transmute_from_u8_to_slice` and `segment::common::mmap_ops::transmute_from_u8_to_mut_slice` ([link](https://github.com/qdrant/qdrant/blob/3fbe1cae6cb7f51a0c5bb4b45cfe6749ac76ed59/lib/common/memory/src/mmap_ops.rs#L106-L118)) call `std::slice::from_raw_parts` with `&[u8]` slice whose start is potentially unaligned for a slice of `&[T]`.

We've observed panics in `segment::index::hnsw_index::graph_links::GraphLinksConverter::serialize_to` — the panic comes from the [precondition assertion in the standard library](https://doc.rust-lang.org/src/core/slice/raw.rs.html#97).

We only observed this when updating from the `nightly-2023-11-01` toolchain to `nightly-2024-03-05`, though the panic also reproduced on various toolchains in between. I believe the issue is still present on older toolchains, but the precondition check was altered slightly in between to make it detect this case. I believe the code is potentially invoking undefined behavior regardless, as the pointer is never guaranteed to be aligned here.

In the specific case of `GraphLinksConverter::serialize_to`, it's relatively easy to fix — the transmuted slice is only used for the convenience of serializing a `&[T]` into a `&[u8]` mmap slice. I have a patch prepared that removes the undefined behavior from this function by transmuting `&[T]` into `&[u8]` and copying that transmuted slice into the `&[u8]` mmap, rather than trying to do the opposite.

However, there are many other usages of these `transmte_from_u8_*` methods throughout the `segment` crate that are vulnerable to the same issue.

## Current Behavior
Panics in `std::slice::from_raw_parts` caused by `GraphLinksConverter::serialize_to`.

<details>
<summary>Backtrace</summary>

```
thread '<unnamed>' panicked at library/core/src/panicking.rs:155:5:
unsafe precondition(s) violated: slice::from_raw_parts_mut requires the pointer to be aligned and non-null, and the total size of the slice not to exceed `isize::MAX`
stack backtrace:
   0:        0x107f100e0 - std::backtrace_rs::backtrace::libunwind::trace::he983ffd396d11726
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/std/src/../../backtrace/src/backtrace/libunwind.rs:104:5
   1:        0x107f100e0 - std::backtrace_rs::backtrace::trace_unsynchronized::h6002dff5d78caefc
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:        0x107f100e0 - std::sys_common::backtrace::_print_fmt::hb1166eaef5f787e8
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/std/src/sys_common/backtrace.rs:68:5
   3:        0x107f100e0 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hba72210463b198ae
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/std/src/sys_common/backtrace.rs:44:22
   4:        0x107f34c48 - core::fmt::rt::Argument::fmt::h8efe6a884a6b4a7c
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/core/src/fmt/rt.rs:142:9
   5:        0x107f34c48 - core::fmt::write::h4cb8d55c8d85b829
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/core/src/fmt/mod.rs:1120:17
   6:        0x107f0c1ac - std::io::Write::write_fmt::h650fcc3e985a9260
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/std/src/io/mod.rs:1846:15
   7:        0x107f0ff20 - std::sys_common::backtrace::_print::h1c7c6a43a3a6a7f9
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/std/src/sys_common/backtrace.rs:47:5
   8:        0x107f0ff20 - std::sys_common::backtrace::print::h70c88364c7e47672
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/std/src/sys_common/backtrace.rs:34:9
   9:        0x107f118a0 - std::panicking::default_hook::{{closure}}::h91c4e952f264ae22
  10:        0x107f11584 - std::panicking::default_hook::h0a763560f5583a48
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/std/src/panicking.rs:292:9
  11:        0x105556180 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h9430742de71445fa
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/alloc/src/boxed.rs:2030:9
  12:        0x105556180 - test::test_main::{{closure}}::heea6bc9506ea3095
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/test/src/lib.rs:138:21
  13:        0x107f121c0 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h05d1fb7a322042c3
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/alloc/src/boxed.rs:2030:9
  14:        0x107f121c0 - std::panicking::rust_panic_with_hook::h4add6ff44d6506d6
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/std/src/panicking.rs:783:13
  15:        0x107f11b54 - std::panicking::begin_panic_handler::{{closure}}::h6b17d83e3f43c72e
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/std/src/panicking.rs:649:13
  16:        0x107f1056c - std::sys_common::backtrace::__rust_end_short_backtrace::he502a3b814db59f8
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/std/src/sys_common/backtrace.rs:171:18
  17:        0x107f118f8 - rust_begin_unwind
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/std/src/panicking.rs:645:5
  18:        0x107f9d628 - core::panicking::panic_nounwind_fmt::runtime::h9b5cd9f9717fd71c
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/core/src/panicking.rs:110:18
  19:        0x107f9d628 - core::panicking::panic_nounwind_fmt::hdf68d2e4a50cf423
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/core/src/panicking.rs:122:9
  20:        0x107f9d6a8 - core::panicking::panic_nounwind::hfc32d53c3917613d
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/core/src/panicking.rs:155:5
  21:        0x105ebbf58 - core::slice::raw::from_raw_parts_mut::precondition_check::h9514179f7a9eac5a
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/core/src/intrinsics.rs:2752:21
  22:        0x105ebbe30 - core::slice::raw::from_raw_parts_mut::h803966888de7facf
                               at /rustc/d18480b84fdbf1efc34f62070951334aa833d761/library/core/src/slice/raw.rs:151:9
  23:        0x105d748d8 - segment::common::mmap_ops::transmute_from_u8_to_mut_slice::hfedc79407c28d87d
                               at /Users/gautam/.cargo/git/checkouts/qdrant-8737e58dc64e4125/76482c6/lib/segment/src/common/mmap_ops.rs:111:14
  24:        0x105f1c150 - segment::index::hnsw_index::graph_links::GraphLinksConverter::serialize_to::had09a0accc44d1f3
                               at /Users/gautam/.cargo/git/checkouts/qdrant-8737e58dc64e4125/76482c6/lib/segment/src/index/hnsw_index/graph_links.rs:237:44
  25:        0x105f1c8a0 - segment::index::hnsw_index::graph_links::GraphLinksConverter::save_as::h680d395a2df29d7a
                               at /Users/gautam/.cargo/git/checkouts/qdrant-8737e58dc64e4125/76482c6/lib/segment/src/index/hnsw_index/graph_links.rs:283:13
  26:        0x105d828f8 - segment::index::hnsw_index::graph_layers_builder::GraphLayersBuilder::into_graph_layers::ha7cfddb3345993af
                               at /Users/gautam/.cargo/git/checkouts/qdrant-8737e58dc64e4125/76482c6/lib/segment/src/index/hnsw_index/graph_layers_builder.rs:88:13
  27:        0x105d2ca78 - <segment::index::hnsw_index::hnsw::HNSWIndex<TGraphLinks> as segment::index::vector_index_base::VectorIndex>::build_index::h2bf2d242ebad7627
                               at /Users/gautam/.cargo/git/checkouts/qdrant-8737e58dc64e4125/76482c6/lib/segment/src/index/hnsw_index/hnsw.rs:637:27
  28:        0x105d89a78 - <segment::index::vector_index_base::VectorIndexEnum as segment::index::vector_index_base::VectorIndex>::build_index::hcee4f9abaf31ca0d
                               at /Users/gautam/.cargo/git/checkouts/qdrant-8737e58dc64e4125/76482c6/lib/segment/src/index/vector_index_base.rs:76:49
  29:        0x105f23c78 - segment::segment_constructor::segment_builder::SegmentBuilder::build::h5e8883e19c35bd52
                               at /Users/gautam/.cargo/git/checkouts/qdrant-8737e58dc64e4125/76482c6/lib/segment/src/segment_constructor/segment_builder.rs:198:17
  30:        0x105b905e4 - vector::qdrant_segments::merge_disk_segments::h08051a9f23100bd0
[snipped, closed-source frames follow]
```

</details>

## Expected Behavior
`std::slice::from_raw_parts` should never be called on an unaligned pointer / should not panic.

## Possible Solution
I can add a patch that would fix the issue for `GraphLinks::serialize_to` specifically, but would not fix all usages of this across the code base. It'd depend on how the transmuted slice is actually needed.

**Comment by generall** [comment]
CC @IvanPleshkov

**Comment by generall** [comment]
Hi @gautamg795! Is there any way to reliably reproduce the panic you observed?

**Comment by gautamg795** [comment]
I don't have a repro yet that doesn't depend on our private code, but will work on trying to minimize. 

Here's the patch that fixes the panic for us, but I'm pretty sure there are other unsafe usages of these transmute methods: https://github.com/qdrant/qdrant/commit/09d6ab9d3817bd3c88f770dd3161f4ec941e0905
Happy to send a PR with this change.

**Comment by generall** [comment]
I am pretty sure we had the discussion about data alignment already, and AFAIK the consensus was that the mmap actually guarantees the alignment of memory, that is why transmuting from raw bytes we considered to be safe. Apparently it is not correct about transmuting in other direction. But I would need some input from my more competent team-members to answer for sure

**Comment by generall** [comment]
Thanks for reporting this btw, @gautamg795 !

**Comment by gautamg795** [comment]
I believe you're correct that `mmap` does return a buffer that's aligned to the largest-possible alignment of the system (i.e. at a page boundary). However, we're slicing into arbitrary positions within the mapped memory to transmute; it's these slices that have no alignment guarantees unless you're very careful about adding padding

**Comment by timvisee** [comment]
Good catch! Thanks for sharing your detailed report.

Correct. Memory maps are aligned, but taking sub slices is problematic. We cannot easily change padding now because we've already had a bunch of releases with this in it.

I like your [`copy_from_slice`](https://doc.rust-lang.org/std/primitive.slice.html#method.copy_from_slice) trick as hack to force align these slices, though I wonder what impact these may have in terms of performance. Though these should be relatively cheap it's obviously more costly than transmuting. We may benefit from a special function that only copies the slice if not aligned. Something like [`align_to`](https://doc.rust-lang.org/std/primitive.slice.html#method.align_to) maybe useful here too.

We could add padding in new versions, in which case it would only be used when writing new vector data from disk. But that wouldn't fix problematic cases we already have on disk due to older versions, and we obviously still have to support those. I do think that padding is the right solution here.

**Comment by gautamg795** [comment]
Thanks for confirming my hypothesis!

I think it actually should be possible to eliminate the intermediate buffers I'm allocating in my example fix -- using either `std::io::Cursor` or just careful arithmetic on the `&[u8]` slice index. I just used the buffers as I didn't totally understand what was happening in the loop, and it was safer this way. 

If it were implemented without the intermediate buffers, I suspect it'd compile to the same instructions as the current code but without any alignment issues.
